apps/src/blockly/addons/cdoSerializationHelpers.js

```diff
@@ -87,57 +81,93 @@ export function positionBlocksOnWorkspace(workspace, blockOrderMap) {
+
+  // If the workspace is RTL, horizontally mirror the starting position
+  let x = RTL ? width - padding : padding;
+  let y = padding;
+
+  let existingColliders = [];
+  let blocksToPlace = [];
+  blocks.forEach(block => {
+    if (isBlockLocationUnset(block)) {
+      blocksToPlace.push(block);
+    } else {
+      insertCollider(existingColliders, getCollider(block));
+    }
+  });
+
+  blocksToPlace.forEach(block => {
+    let collider = getCollider(block);
+    existingColliders.forEach(existingCollider => {
+      if (isOverlapping(collider, existingCollider)) {
+        y =
+          existingCollider.y +
+          existingCollider.height +
+          VERTICAL_SPACE_BETWEEN_BLOCKS +
+          SVG_FRAME_TOP_PADDING;
+        block.moveTo({x, y});
+        // Collision area must be updated to account for new position
+        collider = getCollider(block);
+      }
+    });
+    insertCollider(existingColliders, collider);
   });
 }
 
-/**
- * Use a cursor to position a block on a workspace (if it does not already have a position)
- * @param {Blockly.Block} block - the block to be moved
- * @param {object} cursor - a location for moving a block, if needed
- * @param {number} cursor.x - an x-coordinate for moving a block
- * @param {number} cursor.y - a y-coordinate for moving a block
- * @return {object} the cursor with updated coordinates
- */
-function positionBlockWithCursor(block, cursor) {
-  if (isBlockLocationUnset(block)) {
-    block.moveTo(getNewLocation(block, cursor));
+function getCollider(block) {
+  const position = block.getRelativeToSurfaceXY();
+  const size = block.getHeightWidth();
+
+  const collider = {
+    ...position,
+    ...size,
+  };
+
+  // SVG frames require us to account for additional height and width
+  // TODO: How do we handle RTL here?
+  if (block.functionalSvg_) {
+    collider.x -= SVG_FRAME_SIDE_PADDING;
+    collider.y -= SVG_FRAME_TOP_PADDING;
+    collider.height += SVG_FRAME_HEIGHT;
+    collider.width += SVG_FRAME_SIDE_PADDING * 2;
   }
-  cursor.y += getCursorYAdjustment(block);
-  return cursor;
+
+  return collider;
 }
 
-/**
- * Determines where the current block should be positioned, based on the cursor
- * @param {Blockly.Block} block - the block to be moved
- * @param {object} cursor - a location for moving a block, if needed
- * @param {number} cursor.x - an x-coordinate for moving a block
- * @param {number} cursor.y - a y-coordinate for moving a block
- */
-export function getNewLocation(block, cursor) {
-  const blockHasFrameSvg = !!block.functionalSvg_;
-  const blockTopPadding = blockHasFrameSvg ? SVG_FRAME_TOP_PADDING : 0;
-  const blockSidePadding = blockHasFrameSvg ? SVG_FRAME_SIDE_PADDING : 0;
-  const isRTL = block.workspace.RTL;
-  const newLocation = {
-    x: cursor.x + (isRTL ? -blockSidePadding : blockSidePadding),
-    y: cursor.y + blockTopPadding,
-  };
-  return newLocation;
+function insertCollider(colliders, item) {
+  const sumItem = item.y + item.height;
```

Comment:
Nit: This is indeed a sum, but I wonder if a label like `itemBottom` would be more descriptive?

Reply:
I struggled so hard for a name for this; thank you!